### ISSUE TITLE
smooth_L1_loss_layer Not Implemented Yet

### ISSUE THREAD

**Issue by gplhegde** [rbgirshick/py-faster-rcnn#130]
Hi,
I am trying follow the steps in Readme.md to retrain ZF network on PASCAL VOC dataset using alternating training approach. I am using CPU mode and accordingly modified train_faster_rcnn_alt_opt.py file to set caffe to CPU mode. Network init happens smoothly. However when the solver starts, I get error **"F0327 11:34:19.894506 17762 smooth_L1_loss_layer.cpp:54] Not Implemented Yet"**.  Why smooth_L1_loss layer not implemented in caffe? Am I using older version? I cloned this repo 2 days back. Attached is the log
Thanks.
[faster_rcnn_alt_opt_ZF_.txt.2016-03-27_11-34-14.txt](https://github.com/rbgirshick/py-faster-rcnn/files/190769/faster_rcnn_alt_opt_ZF_.txt.2016-03-27_11-34-14.txt)

**Comment by happyharrycn** [comment]
The forward/backward method on smooth_L1_loss_layer.cpp is not implemented on CPU. It makes sense since training on CPU is almost two magnitude slower and thus becoming infeasible for even a small dataset.

**Comment by gplhegde** [comment]
I see. Thanks.

**Comment by lokhandeomkar** [comment]
@gplhegde , @happyharrycn so this example cannot be run on a CPU right? Is there an example for training and testing py-faster-rcnn on your own dataset with CPU mode that you guys know of?

**Comment by gplhegde** [comment]
You can use CPU only mode for testing. You need GPU card with more than 5GB memory for training. I could not train the smallest supported network in my 2GB card.

**Comment by Georgezhouzhou** [comment]
_**I tried to implement the CPU mode, but this still can not pass the runtest. Hope it helps, :)**_

`// ------------------------------------------------------------------
// Fast R-CNN
// Copyright (c) 2015 Microsoft
// Licensed under The MIT License [see fast-rcnn/LICENSE for details]
// Written by Ross Girshick
// ------------------------------------------------------------------

#include "caffe/fast_rcnn_layers.hpp"

namespace caffe {

template <typename Dtype>
void SmoothL1LossLayer<Dtype>::LayerSetUp(
  const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
  SmoothL1LossParameter loss_param = this->layer_param_.smooth_l1_loss_param();
  sigma2_ = loss_param.sigma() * loss_param.sigma();
  has_weights_ = (bottom.size() >= 3);
  if (has_weights_) {
    CHECK_EQ(bottom.size(), 4) << "If weights are used, must specify both "
      "inside and outside weights";
  }
}

template <typename Dtype>
void SmoothL1LossLayer<Dtype>::Reshape(
  const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
  LossLayer<Dtype>::Reshape(bottom, top);
  CHECK_EQ(bottom[0]->channels(), bottom[1]->channels());
  CHECK_EQ(bottom[0]->height(), bottom[1]->height());
  CHECK_EQ(bottom[0]->width(), bottom[1]->width());
  if (has_weights_) {
    CHECK_EQ(bottom[0]->channels(), bottom[2]->channels());
    CHECK_EQ(bottom[0]->height(), bottom[2]->height());
    CHECK_EQ(bottom[0]->width(), bottom[2]->width());
    CHECK_EQ(bottom[0]->channels(), bottom[3]->channels());
    CHECK_EQ(bottom[0]->height(), bottom[3]->height());
    CHECK_EQ(bottom[0]->width(), bottom[3]->width());
  }
  diff_.Reshape(bottom[0]->num(), bottom[0]->channels(),
      bottom[0]->height(), bottom[0]->width());
  errors_.Reshape(bottom[0]->num(), bottom[0]->channels(),
      bottom[0]->height(), bottom[0]->width());
  // vector of ones used to sum
  ones_.Reshape(bottom[0]->num(), bottom[0]->channels(),
      bottom[0]->height(), bottom[0]->width());
  for (int i = 0; i < bottom[0]->count(); ++i) {
    ones_.mutable_cpu_data()[i] = Dtype(1);
  }
}


template <typename Dtype>
void SmoothL1LossLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
    const vector<Blob<Dtype>*>& top) {
  // NOT_IMPLEMENTED; 
  int count = bottom[0]->count();
  //int num = bottom[0]->num();
  const Dtype* in = diff_.cpu_data();
  Dtype* out = errors_.mutable_cpu_data();
  caffe_set(errors_.count(), Dtype(0), out);

  caffe_sub(
      count,
      bottom[0]->cpu_data(),
      bottom[1]->cpu_data(),
      diff_.mutable_cpu_data());    // d := b0 - b1
  if (has_weights_) {
    // apply "inside" weights
    caffe_mul(
        count,
        bottom[2]->cpu_data(),
        diff_.cpu_data(),
        diff_.mutable_cpu_data());  // d := w_in * (b0 - b1)
  }


 
 for (int index = 0;index < count; ++index){
    Dtype val = in[index];
    Dtype abs_val = abs(val);
    if (abs_val < 1.0 / sigma2_) {
      out[index] = 0.5 * val * val * sigma2_;
    } else {
      out[index] = abs_val - 0.5 / sigma2_;
    }
  }

  if (has_weights_) {
    // apply "outside" weights
    caffe_mul(
        count,
        bottom[3]->cpu_data(),
        errors_.cpu_data(),
        errors_.mutable_cpu_data());  // d := w_out * SmoothL1(w_in * (b0 - b1))
  }

  Dtype loss = caffe_cpu_dot(count, ones_.cpu_data(), errors_.cpu_data());
  top[0]->mutable_cpu_data()[0] = loss / bottom[0]->num();
}

template <typename Dtype>
void SmoothL1LossLayer<Dtype>::Backward_cpu(const vector<Blob<Dtype>*>& top,
    const vector<bool>& propagate_down, const vector<Blob<Dtype>*>& bottom) {
  // NOT_IMPLEMENTED; 
  int count = diff_.count();
  //int num = diff_.num();
  const Dtype* in = diff_.cpu_data();
  Dtype* out = errors_.mutable_cpu_data();
  caffe_set(errors_.count(), Dtype(0), out);
  
  for (int index = 0;index < count; ++index){
    Dtype val = in[index];
    Dtype abs_val = abs(val);
    if (abs_val < 1.0 / sigma2_) {
      out[index] = sigma2_ * val;
    } else {
      out[index] = (Dtype(0) < val) - (val < Dtype(0));
    }
  }

  for (int i = 0; i < 2; ++i) {
    if (propagate_down[i]) {
      const Dtype sign = (i == 0) ? 1 : -1;
      const Dtype alpha = sign * top[0]->cpu_diff()[0] / bottom[i]->num();
      caffe_cpu_axpby(
          count,                           // count
          alpha,                           // alpha
          diff_.cpu_data(),                // x
          Dtype(0),                        // beta
          bottom[i]->mutable_cpu_diff());  // y
      if (has_weights_) {
        // Scale by "inside" weight
        caffe_mul(
            count,
            bottom[2]->cpu_data(),
            bottom[i]->cpu_diff(),
            bottom[i]->mutable_cpu_diff());
        // Scale by "outside" weight
        caffe_mul(
            count,
            bottom[3]->cpu_data(),
            bottom[i]->cpu_diff(),
            bottom[i]->mutable_cpu_diff());
      }
    }
  }
}

#ifdef CPU_ONLY
STUB_GPU(SmoothL1LossLayer);
#endif

INSTANTIATE_CLASS(SmoothL1LossLayer);
REGISTER_LAYER_CLASS(SmoothL1Loss);

}  // namespace caffe
`

**Comment by Lukeeeeee** [comment]
@zhouphd   I also find this CPU implementation https://github.com/patricio-astudillo/caffe-fast-rcnn/commit/0e4f8c8108533d00b913d0422448fd166f5bf2e5 
I replace older one **fast-rcnn/src/caffe/layers/smooth_L1_loss_layer.cpp** with the layer above 
and still can not pass the runtest. 
Don't know why.

**Comment by zhouphd** [comment]
@Lukeeeeee This may be the double & float issue, please watch this branch, https://github.com/Austriker/caffe/pull/7 

Strongly suggest you to get a GPU to run py-faster-rcnn, good luck

**Comment by MarcosPieras** [comment]
HI, 
@Lukeeeeee , did you solve it ?

**Comment by Lukeeeeee** [comment]
Sorry, I did not.
 I try to solve it by using  some cpu implementations of other
contributors, but it did not work. So I just gave up. So good luck, wish u
can sovle this.

**Comment by becomedragon** [comment]
https://github.com/rbgirshick/caffe-fast-rcnn/pull/17/commits/6309fd7ebab9b4140b41974a7f10fd1fa540d9d3

this is the implemented edition.
replace these file,and make -j8 again , can solve the problem

**Comment by PeterJiangwy** [comment]
@becomedragon, I follow these steps, but it still not work, could you help me.

**Comment by PeterJiangwy** [comment]
@zhouphd, I met the same problem, "smooth_L1_loss_layer.cpp:54] Not Implemented Yet", could you give me some suggestions?

**Comment by trunghoangminh** [comment]
@zhouphd I also follow by your step. But It still encounters this issue?

**Comment by trunghoangminh** [comment]
@PeterJiangwy, I followed by this post http://blog.csdn.net/qq_14975217/article/details/51495844. The issue is solved

**Comment by cobbwho** [comment]
@trunghoangminh  I am modifying the file according to the blog you mentioned, The error of 'Not implemented Yet' has disappeared, but the popup dialog says python stopped working. Have you encountered this problem? By the way, I'm using win10. Thanks any way.
